feat(edit-content): add system timezone support to calendar field component

oidacra opened the pull request:
This pull request introduces timezone support for calendar fields and refactors related components to improve functionality and maintainability. Key updates include adding a system timezone configuration, enhancing the `DotEditContentCalendarFieldComponent` with timezone handling, and updating related utilities and interfaces to support the changes.

### Timezone Support Enhancements:
* Added a `SystemTimezone` interface and integrated it into `ConfigParams` in `dotcms-config.service.ts` to provide system timezone configuration. Updated the `DotcmsConfigService` to include the new `systemTimezone` property. [[1]](diffhunk://#diff-9d02a07b0759480653f520aaeb3df576768ab3d445afbff8db9da1a24eb0e800R30-R35) [[2]](diffhunk://#diff-9d02a07b0759480653f520aaeb3df576768ab3d445afbff8db9da1a24eb0e800R57) [[3]](diffhunk://#diff-9d02a07b0759480653f520aaeb3df576768ab3d445afbff8db9da1a24eb0e800L115-R125)
* Updated `DotEditContentCalendarFieldComponent` to handle timezone adjustments, including computing default dates and displaying timezone labels when applicable. Added utility methods for timezone conversion and adjusted the component's structure for better readability. [[1]](diffhunk://#diff-45105f940f5c94adeb34f9efd7a3888ebb8d90e1c76dfd1d5ee2fd457ba2d817L1-R36)R1, [[2]](diffhunk://#diff-45105f940f5c94adeb34f9efd7a3888ebb8d90e1c76dfd1d5ee2fd457ba2d817L25-R163) [[3]](diffhunk://#diff-7b53b3487cc14a70194752d39ec752b0225874d182fc78b0f19e26b036dd7d48R1-R43) [[4]](diffhunk://#diff-d3cf1dcaf1f9ab16422a2053e6eb9cc42c97e3c3f4101504ab46bd6664d66a47R1-R7)

### Component Refactoring:
* Modified `DotEditContentFieldComponent` and `DotEditContentFormComponent` to inject `DotcmsConfigService` and compute the system timezone as a reactive signal. Updated templates to pass the timezone to child components. [[1]](diffhunk://#diff-f03434ad88316542a23c3fb0bf50b6650130632c3e7953d7ee3c882cb5950d06R10-R16) [[2]](diffhunk://#diff-f03434ad88316542a23c3fb0bf50b6650130632c3e7953d7ee3c882cb5950d06L74-R82) [[3]](diffhunk://#diff-f03434ad88316542a23c3fb0bf50b6650130632c3e7953d7ee3c882cb5950d06R100-R112) [[4]](diffhunk://#diff-52fdc67adc70be54cf6c07ee18c9340c26cbf50a4d8fe5ed14a21d33def30a23L13-R13) [[5]](diffhunk://#diff-52fdc67adc70be54cf6c07ee18c9340c26cbf50a4d8fe5ed14a21d33def30a23R36) [[6]](diffhunk://#diff-52fdc67adc70be54cf6c07ee18c9340c26cbf50a4d8fe5ed14a21d33def30a23R122-R124) [[7]](diffhunk://#diff-52fdc67adc70be54cf6c07ee18c9340c26cbf50a4d8fe5ed14a21d33def30a23R180-R184)
* Updated `dot-edit-content-calendar-field.component.html` to include hints and timezone labels dynamically based on the system timezone.

### Utility and Interface Updates:
* Added `CALENDAR_FIELD_TYPES_WITH_TIME` constant to distinguish calendar field types that include time. Updated related utilities and constants accordingly. [[1]](diffhunk://#diff-8bf2337f0dcafc78712e3ef3394b00a1cccdb0df916df36309dff5a40465e7f9R12-R13) [[2]](diffhunk://#diff-b83381c24842c1c58b78cd24259b3eb04632858f0ac2f4079d66a1dccf97a2edL4-R14) [[3]](diffhunk://#diff-99ac6732d97f8e73c67b5464dfb645a4ae4ec51a80005cf22d7f83a27dc43c74L14-R14) [[4]](diffhunk://#diff-99ac6732d97f8e73c67b5464dfb645a4ae4ec51a80005cf22d7f83a27dc43c74L87-R99)
* Extended `FormFieldValue` type to support `number` for better flexibility in form data handling.

These changes collectively enhance the system's ability to handle timezones effectively while maintaining a clean and modular codebase.
### Checklist
- [ ] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

This PR fixes: #31248

oidacra committed:
Implement system configuration service and update date handling

- Introduced  to manage system configuration data from the DotCMS server.
- Added new models for system configuration, including , , and related interfaces.
- Updated date handling in  to utilize  for timezone-aware date formatting.
- Enhanced calendar field components to support timezone conversion and improved date handling.
- Refactored existing components and services to integrate the new system configuration and date handling logic.
- Added unit tests for the new service and updated existing tests to cover new functionality.

This commit lays the groundwork for improved date management and system configuration access across the application.

oidacra committed:
Refactor DotEditContentCalendarField to simplify date handling

- Updated the DotEditContentCalendarField component to restrict the lastUtcValue and onChange callback to only accept Date or null types, removing unnecessary types for improved type safety.
- Adjusted writeValue and registerOnChange methods accordingly to enhance clarity and maintainability.

oidacra committed:
Enhance DotFormatDateService and DotEditContentCalendarField for improved date handling

- Refactored DotFormatDateService to streamline locale loading and fallback mechanisms for date formatting.
- Updated DotEditContentCalendarField to utilize new date handling logic, ensuring better timezone management and field configuration.
- Introduced new properties for system timezone and content type in DotEditContentCalendarField to enhance flexibility.
- Improved unit tests to cover new functionalities and ensure robust date handling across components.

zJaaal commented on core-web/libs/dotcms-models/src/lib/dot-system-config.model.ts:
```diff
@@ -0,0 +1,82 @@
+        // Other config properties we don't need for the system config
+        [key: string]: unknown;
+    };
+    menu?: unknown; // Menu data from the endpoint
```
I think we have a type for this

nicobytes commented on core-web/package.json:
```diff
@@ -103,8 +104,8 @@
-        "date-fns": "^2.29.3",
-        "date-fns-tz": "^1.1.6",
+        "date-fns": "^4.0.0",
+        "date-fns-tz": "^3.2.0",
```
Could we remove this dep?

oidacra commented on yarn.lock:
No. Because I updated 2 libs.

oidacra commented on core-web/libs/edit-content/src/lib/fields/dot-edit-content-calendar-field/dot-edit-content-calendar-field.component.ts:
```diff
@@ -1,27 +1,245 @@
+
+                // Use setTimeout to ensure onChange callback is fully initialized by Angular Forms
+                // This is necessary because writeValue can be called before registerOnChange
+                // which is a known timing issue with ControlValueAccessor initialization
```
removed. I forgot clean this one.

oidacra committed:
Refactor DotSystemConfig and DotEditContentCalendarField for improved type safety and styling

oidacra committed:
Update date handling in DotEditContent components for improved timestamp processing

- Upgraded the @date-fns/tz package to version 1.4.0 for enhanced timezone support.
- Refactored date resolution functions to ensure consistent handling of numeric timestamps and string inputs.
- Updated DotEditContentCalendarField to store timestamps as numbers, improving type safety and reducing potential errors.
- Enhanced utility functions to process calendar field values, ensuring they return numeric timestamps or null as appropriate.
- Improved unit tests to cover new date handling logic and ensure robust functionality across components.

oidacra committed:
Enhance unit tests for DotSiteSelector and DotPersonaSelector components with mock system configuration

oidacra committed:
Enhance unit tests by integrating MockDotSystemConfigService across multiple components

- Added MockDotSystemConfigService to provide mock system configuration for unit tests in DotTemplateCreateEdit, DotContentTypeCopyDialog, DotCreatePersonaForm, and DOTTestBed.
- Updated relevant test files to utilize the mock service, ensuring consistent system configuration access during testing.
- Improved test coverage for components relying on system configuration data.

oidacra committed:
chore: update yarn.lock

oidacra committed:
fix: update unit tests for DotContentDriveShellComponent to include DotSystemConfigService

oidacra committed:
Merge branch 'main' into 31248-UTM-label
